src/js/trove/load-lib.js

```diff
@@ -310,31 +345,48 @@
         }
       };
 
-
-      return runtime.pauseStack(function(restarter) {
+      var pauseFunction = function(restarter) {
         var mainReached = false;
         var mainResult = "Main result unset: should not happen";
         postLoadHooks[main] = function(answer) {
           mainReached = true;
           mainResult = answer;
-        }
-        return otherRuntime.runThunk(function() {
+        };
+
+        var thunk = function() {
           otherRuntime.modules = realm;
           return otherRuntime.runStandalone(staticModules, realm, depMap, toLoad, postLoadHooks);
-        }, function(result) {
+        };
+
+        var then = function(result) {
+          var retVal;
+          
           if(!mainReached) {
             // NOTE(joe): we should only reach here if there was an error earlier
             // on in the chain of loading that stopped main from running
-            restarter.resume(makeModuleResult(otherRuntime, result, makeRealm(realm), runtime.nothing));
+            retVal = makeModuleResult(otherRuntime, result, makeRealm(realm), runtime.nothing);
           }
           else {
             var finalResult = otherRuntime.makeSuccessResult(mainResult);
             finalResult.stats = result.stats;
-            restarter.resume(makeModuleResult(otherRuntime, finalResult, makeRealm(realm), runtime.nothing));
+            retVal = makeModuleResult(otherRuntime, finalResult, makeRealm(realm), runtime.nothing);
           }
-        });
-      });
 
+          if (runtime.bounceAllowed) {
+            restarter.resume(retVal);
+          } else {
+            return retVal;
+          }
```

Comment:
This pattern seems to come up a lot; I wonder if there could be a runtime method or a method on the restarter that abstracts this away so we can just write `return restarter.resumeOrReturn(retVal);`

In the call to `pauseFunction`, maybe we should pass in an argument to `pauseFunction` that has a dummy restarter?

It seems like there's a pattern here to abstract one way or another.

Another idea is to have `runtime.maybePauseStack` which simply calls the callback with the right dummy restarter if `bounceAllowed` is false, or does the normal pausing flow otherwise.  If a part of the program _will not work_ without pausing, it could use `runtime.pauseStack`, which will throw an error if `bounceAllowed` is false.